src/__tests__/unit/controllers/hooks/router-store/test.tsx

```diff
@@ -0,0 +1,306 @@
+  },
+  {
+    path: '/blah',
+    component: () => <div>path</div>,
+    name: '',
+  },
+];
+
+const historyBuildOptions = {
+  initialEntries: [
+    `${mockLocation.pathname}${mockLocation.search}${mockLocation.hash}`,
+  ],
+};
+
+let history = historyHelper.createMemoryHistory(historyBuildOptions);
+let historyPushSpy = jest.spyOn(history, 'push');
+const nextTick = () => new Promise(resolve => setTimeout(resolve));
+
+const MockComponent = ({ children, ...rest }: any) => {
+  return children(rest);
+};
+
+describe('useQueryParam', () => {
+  beforeEach(() => {
+    history = historyHelper.createMemoryHistory(historyBuildOptions);
+    historyPushSpy = jest.spyOn(history, 'push');
+  });
+
+  afterEach(() => {
+    defaultRegistry.stores.clear();
+    jest.resetAllMocks();
+    jest.restoreAllMocks();
+  });
+
+  it('should return the right param value', () => {
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param] = useQueryParam('foo');
+            expect(param).toEqual('hello');
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+  });
+
+  it('should return undefined for non-existent params', () => {
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param] = useQueryParam('iamnotaqueryparam');
+            expect(param).toEqual(undefined);
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+  });
+
+  it('should update URL with new param value', async () => {
+    const mockPath = mockLocation.pathname;
+    let qpVal: string | undefined, qpUpdateFn: (qp: string) => void;
+
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param, setParam] = useQueryParam('foo');
+            qpVal = param;
+            qpUpdateFn = setParam;
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+
+    expect(qpVal).toEqual('hello');
+
+    act(() => qpUpdateFn('newVal'));
+
+    await nextTick();
+
+    expect(historyPushSpy).toBeCalledWith(
+      `${mockPath}?foo=newVal&bar=world#hash`
+    );
+  });
+
+  it('should remove param from URL when set to null', async () => {
+    const mockPath = mockLocation.pathname;
+    let qpVal: string | undefined, qpUpdateFn: (qp: string | null) => void;
+
+    mount(
+      <Router routes={mockRoutes} history={history}>
+        <MockComponent>
+          {() => {
+            const [param, setParam] = useQueryParam('foo');
+            qpVal = param;
+            qpUpdateFn = setParam;
+
+            return <div>I am a subscriber</div>;
+          }}
+        </MockComponent>
+      </Router>
+    );
+
+    expect(qpVal).toEqual('hello');
+
+    act(() => qpUpdateFn(null));
```

Comment:
I might be overthinking this, but is removing the querystring parameter the desired behavior?

I put together a quick sandbox, using `qs.stringify` and `history.push`. In the example I call `qs.stringify` with `{ foo: null}` and the path is updated to `/home?foo=` which seems easier to reason about.

You can see the sandbox here https://codesandbox.io/s/react-resource-router-basic-routing-example-forked-pyg2m?file=/src/home/index.tsx

Reply:
This is something I came up with during API design. The motivation behind the decision to be able to remove the param from the URL is —
1. To be able to differentiate between empty value and non-existent value - `/path?foo=` vs `/path` where foo is non-existent as opposed to "we don't know if it's non-existent or empty".
2. I created a forked example of the codesandbox you added - https://codesandbox.io/s/react-resource-router-query-string-exp-forked-dxfcs?file=/src/home/index.tsx. You can see the output URL of both the sandboxes is the same `/about?foo=` but the value we set in your sandbox is null and the value we set in my sandbox is an empty string `''`. With the API design we created, users can choose to either pass an empty string to just remove the value or pass undefined/null to remove the param altogether.

Thoughts?

Reply:
I like @flexdinesh approach better, otherwise consumers cannot remove query parameters (and sometimes in Jira they have to)

Reply:
Cool that simplifies it. Nice one @flexdinesh !